src/util/taproot.rs

```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+                // Needs an explicit push to add the None that we just popped.
+                // Cannot use .last() because of borrow checker issues.
+                Some(None) => {
+                    self.branch.push(None);
+                    break;
+                } // Cannot combine further
+            };
+            if depth == 0 {
+                // We are trying to combine two nodes at root level.
+                // Can't propagate further up than the root
+                return Err(TaprootBuilderError::OverCompleteTree);
+            }
+            node = NodeInfo::combine(node, child)?;
+            // Propagate to combine nodes at a lower depth
+            depth -= 1;
+        }
+
+        if self.branch.len() < depth + 1 {
+            // add enough nodes so that we can insert node at depth `depth`
+            let num_extra_nodes = depth + 1 - self.branch.len();
+            self.branch
+                .extend((0..num_extra_nodes).into_iter().map(|_| None));
+        }
+        // Push the last node to the branch
+        self.branch[depth] = Some(node);
+        Ok(self)
+    }
+}
+
+// Internally used structure to represent the node information in taproot tree
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct NodeInfo {
+    /// Merkle Hash for this node
+    hash: sha256::Hash,
+    /// information about leaves inside this node
+    leaves: Vec<LeafInfo>,
+}
+
+impl NodeInfo {
+    // Create a new NodeInfo with omitted/hidden info
+    fn new_hidden(hash: sha256::Hash) -> Self {
+        Self {
+            hash: hash,
+            leaves: vec![],
+        }
+    }
+
+    // Create a new leaf with NodeInfo
+    fn new_leaf_with_ver(script: Script, ver: LeafVersion) -> Self {
+        let leaf = LeafInfo::new(script, ver);
+        Self {
+            hash: leaf.hash(),
+            leaves: vec![leaf],
+        }
+    }
+
+    // Combine two NodeInfo's to create a new parent
+    fn combine(a: Self, b: Self) -> Result<Self, TaprootBuilderError> {
+        let mut all_leaves = Vec::with_capacity(a.leaves.len() + b.leaves.len());
+        for mut a_leaf in a.leaves {
+            a_leaf.merkle_branch.push(b.hash)?; // add hashing partner
+            all_leaves.push(a_leaf);
+        }
+        for mut b_leaf in b.leaves {
+            b_leaf.merkle_branch.push(a.hash)?; // add hashing partner
+            all_leaves.push(b_leaf);
+        }
+        let mut eng = TapBranchHash::engine();
+        if a.hash < b.hash {
+            eng.input(&a.hash);
+            eng.input(&b.hash);
+        } else {
+            eng.input(&b.hash);
+            eng.input(&a.hash);
+        };
+        Ok(Self {
+            hash: sha256::Hash::from_engine(eng),
+            leaves: all_leaves,
+        })
+    }
+}
+
+// Internally used structure to store information about taproot leaf node
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct LeafInfo {
+    /// The underlying script
```

Comment:
nit: we should either use doc-style or normal comments at both struct and fields

Reply:
Normal comments because its a private struct, right?

Reply:
I believe the point here is consistency and I prefer using doc comments. They will be only visible if doc is generated with special flags.

Reply:
Oh wow, I have been doing patch review for the last two years telling folks not to use `///` for private structs based on the docs build systems in other languages. Thanks for schooling me.